# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/u2f/U2F.h
# Commit: f7937d3264db
# Full Hash: f7937d3264db00771b46cb1fcba71640d8df05cb
# Author: J.C. Jones <jjones@mozilla.com>
# Date: 2019-03-30 09:32:04
# Regressor Bug: 1448408
# File Overlap Count: 2
# Description:
#   Bug 1448408 - Web Authentication - Don't immediately abort on visibility events r=keeler
#   
#   The published recommendation of L1 for WebAuthn changed the visibility/focus
#   listening behaviors to a SHOULD [1], and Chromium, for reasons like our SoftU2F
#   bug [0], opted to not interrupt on tabswitch/visibility change.
# ==============================================================================

diff -r a5c4f7e6069a -r f7937d3264db dom/u2f/U2F.h
--- a/dom/u2f/U2F.h	Fri Mar 29 19:38:53 2019 +0000
+++ b/dom/u2f/U2F.h	Fri Mar 29 17:59:08 2019 +0000
@@ -40,7 +40,9 @@
 
  public:
   explicit U2FTransaction(const U2FCallback&& aCallback)
-      : mCallback(std::move(aCallback)), mId(NextId()) {
+      : mCallback(std::move(aCallback)),
+        mId(NextId()),
+        mVisibilityChanged(false) {
     MOZ_ASSERT(mId > 0);
   }
 
@@ -66,6 +68,10 @@
   // Unique transaction id.
   uint64_t mId;
 
+  // Whether or not visibility has changed for the window during this
+  // transaction
+  bool mVisibilityChanged;
+
  private:
   // Generates a unique id for new transactions. This doesn't have to be unique
   // forever, it's sufficient to differentiate between temporally close
@@ -124,7 +130,9 @@
  protected:
   // Cancels the current transaction (by sending a Cancel message to the
   // parent) and rejects it by calling RejectTransaction().
-  MOZ_CAN_RUN_SCRIPT void CancelTransaction(const nsresult& aError) override;
+  MOZ_CAN_RUN_SCRIPT void CancelTransaction(const nsresult& aError);
+  // Upon a visibility change, makes note of it in the current transaction.
+  MOZ_CAN_RUN_SCRIPT void HandleVisibilityChange() override;
 
  private:
   MOZ_CAN_RUN_SCRIPT ~U2F();